### ISSUE TITLE
Failed to restart daemon after 24.0.2 -> 24.0.3 upgrade

### ISSUE THREAD

**Issue by kotso** [moby/moby#45898]
### Description

After update on all hosts (20+, all that I'm running) docker daemon is failed.

error is:

```
Jul 07 03:03:17 server systemd[1]: Starting Docker Application Container Engine...
Jul 07 03:03:17 server dockerd[52024]: time="2023-07-07T03:03:17.076517762Z" level=info msg="Starting up"
Jul 07 03:03:17 server dockerd[52024]: time="2023-07-07T03:03:17.086137356Z" level=info msg="[graphdriver] using prior storage driver: fuse-overlayfs"
Jul 07 03:03:17 server dockerd[52024]: time="2023-07-07T03:03:17.103391077Z" level=info msg="Loading containers: start."
Jul 07 03:03:17 server dockerd[52024]: time="2023-07-07T03:03:17.411734779Z" level=info msg="there are running containers, updated network configuration will not take affect"
Jul 07 03:03:17 server dockerd[52024]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 07 03:03:17 server dockerd[52024]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x564e9be4c7c0]
Jul 07 03:03:17 server dockerd[52024]: goroutine 607 [running]:
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).prepareMountPoints(0xc00054b7b8?, 0xc0007c2500)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/mounts.go:24 +0x1c0
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).restore.func5(0xc0007c2500, 0x0?)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:552 +0x271
Jul 07 03:03:17 server dockerd[52024]: created by github.com/docker/docker/daemon.(*Daemon).restore
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:530 +0x8d8
Jul 07 03:03:17 server dockerd[52024]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 07 03:03:17 server dockerd[52024]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x564e9be4c7c0]
Jul 07 03:03:17 server dockerd[52024]: goroutine 601 [running]:
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).prepareMountPoints(0xc00054b7b8?, 0xc000cb2000)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/mounts.go:24 +0x1c0
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).restore.func5(0xc000cb2000, 0x0?)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:552 +0x271
Jul 07 03:03:17 server dockerd[52024]: created by github.com/docker/docker/daemon.(*Daemon).restore
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:530 +0x8d8
Jul 07 03:03:17 server dockerd[52024]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 07 03:03:17 server dockerd[52024]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x564e9be4c7c0]
Jul 07 03:03:17 server dockerd[52024]: goroutine 603 [running]:
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).prepareMountPoints(0xc00054b7b8?, 0xc000187680)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/mounts.go:24 +0x1c0
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).restore.func5(0xc000187680, 0x0?)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:552 +0x271
Jul 07 03:03:17 server dockerd[52024]: created by github.com/docker/docker/daemon.(*Daemon).restore
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:530 +0x8d8
Jul 07 03:03:17 server dockerd[52024]: panic: runtime error: invalid memory address or nil pointer dereference
Jul 07 03:03:17 server dockerd[52024]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x564e9be4c7c0]
Jul 07 03:03:17 server dockerd[52024]: goroutine 600 [running]:
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).prepareMountPoints(0xc00054b7b8?, 0xc00063e500)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/mounts.go:24 +0x1c0
Jul 07 03:03:17 server dockerd[52024]: github.com/docker/docker/daemon.(*Daemon).restore.func5(0xc00063e500, 0xc00005e600?)
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:552 +0x271
Jul 07 03:03:17 server dockerd[52024]: created by github.com/docker/docker/daemon.(*Daemon).restore
Jul 07 03:03:17 server dockerd[52024]:         /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:530 +0x8d8
```

I'm running daemon with config:

```
{
  "live-restore": true
}
```

any chance to manage restart daemon without restarting containers?

### Reproduce

systemctl restart docker

### Expected behavior

systemd service to start

### docker version

```bash
unable to provide, as service is down
```

### docker info

```bash
unable to provide, as service is down
```

### Additional Info

cant provide docker info as service do not start, I have containers running on servers.

If i reboot server (i.e. containers get stopped, docker starts without issues)

**Comment by calvinbui** [comment]
the panic is caused by scripts updating running container network configs. I worked around this:

1. rolling back to the previous version: `sudo apt-get install docker-ce="5:24.0.2-1~ubuntu.22.04~jammy"`
2. stopping all containers: `docker stop $(docker ps -a -q)`
3. perform the upgrade to docker
4. start all containers: `docker start $(docker ps -a -q)`

**Comment by tomsiewert** [comment]
We used `ctr -n moby task ls -q | xargs ctr -n moby task kill ; apt install -f` to recover the already broken systems without rebooting them.

**Comment by thaJeztah** [comment]
- likely suspect; https://github.com/moby/moby/pull/45754

**Comment by ap-wtioit** [comment]
This happened to me already on 3 systems (have 2 more to offer for debugging before i have no test envs left).

Do you need any more info?

**Comment by ap-wtioit** [comment]
It seems after rolling back the update with `sudo apt-get install docker-ce="5:24.0.2-1~ubuntu.22.04~jammy"
` (from @calvinbui https://github.com/moby/moby/issues/45898#issuecomment-1624779020) our server (that had the pending update) is working fine again without the need to stop or remove the containers.

On ubuntu/debian the previously installed version can be found with ` sudo apt-cache madison docker-ce`.

**Comment by neersighted** [comment]
Debug logs, as well as information on what kinds of volumes you have mounted would be helpful. Are you making use of any Volume plugins or Swarm CSI drivers, or are they all 'local' volumes?

**Comment by ap-wtioit** [comment]
No Swarm CSI drivers, no volume drivers (afaik):
`docker volume ls`
```
local     3b88079ba2faf2c155d2c1e211d46a48993fb13aba7ff1979a4e0c137a624045
local     4a9635073be636d8aa3cc015252610aba85a8a8317e7016c9cf51c8d485ff448
local     5af16cb82ac9697f48d73a3ea8606fd988744d7202e62766b0dfbdae09134398
local     8f37325f4109dcd617dd3e31226b3b55e07a34082f37eb45dc5aa78d20fec914
local     29b87479a68961c2ca2efc420fffc4ab6cd826634619ca1e498cc867deca3e95
local     666c829cbb9eb9b888f165568be3f0b0fae4a2e63239218a52df2ca3948ac8b0
local     48917ae2929ba20e74e586db7862e612081c800efcf18595470f2dae6b45e44c
local     a0949226237fcf2945e9e60d0e0637abecc080803b0032c094d9b0e9ec495793
local     af213695e4c4c618d5e836ddfe364a4aefe9777773d7843dd1a77514ef2ca9d2
local     c23da8331dc4967a4fc837c909f0727264663ff1764f44463fae2ad4277ee4f4
local     c26ea74766cb1984ae04b5aa60ea073543dc699534fa2decab39c49e1224e4b4
local     daf407dccef38ef9cc6bb4e21dbe06d5a0302ad88b3d26416bc5299113a5c3b0
local     redacted_name
local     redacted_name2
local     f1ef13a0fbe7c672c3dfbcb67b0c85ab93594a282fc073a8422e66e8988f273a
local     ff6b60f57037e60eeb520e67a745b692a629ee5215bc3ab483cec6803403adf1
local     runner-bdn9e36m-project-20-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-20-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-118-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-118-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-148-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-148-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-221-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-221-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-222-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-231-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-233-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-254-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-254-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-271-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-271-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-288-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-350-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-350-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-387-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-422-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-422-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-437-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-437-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-447-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-447-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-462-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-497-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-497-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-530-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-530-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-532-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-532-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-535-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-535-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-548-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-548-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-550-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
local     runner-bdn9e36m-project-550-concurrent-0-cache-c33bcaa1fd2c77edfc3893b41966cea8
local     runner-bdn9e36m-project-558-concurrent-0-cache-3c3f060a0374fc8bc39395164f415a70
```
(No (gitlab) runner caches where present on my local workstation only local volumes as well)

**Comment by thaJeztah** [comment]
I think the problem is here; https://github.com/moby/moby/blob/1d9c8619cded4657af1529779c5771127e8ad0e7/daemon/volumes.go#L243-L252C2


That function only looks at the volume if it has Driver set _and_ Volume is `nil`, but if there's no `Driver`, it doesn't bother with the Volume, and returns nil.

And because there's no error, the code continues with the branch below that function call, which includes printing a log message;
https://github.com/moby/moby/blob/1d9c8619cded4657af1529779c5771127e8ad0e7/daemon/mounts.go#L18-L25

**Comment by thaJeztah** [comment]
- opened https://github.com/moby/moby/pull/45902

**Comment by thaJeztah** [comment]
> Edit: no, the driver for local volumes is local; a nil driver seems to signify 'unknown, try all possible drivers.'

It's empty for bind-mounts etc, which don't need restoring

**Comment by thaJeztah** [comment]
See https://github.com/moby/moby/blob/1d9c8619cded4657af1529779c5771127e8ad0e7/volume/mounts/mounts.go#L170-L174

**Comment by kotso** [comment]
Issue fixed with 24.0.4 update.
